packages/strapi-generate-new/lib/merge-template.js

```diff
@@ -0,0 +1,164 @@
+'use strict';
+
+const path = require('path');
+const fse = require('fs-extra');
+const fetch = require('node-fetch');
+const unzip = require('unzip-stream');
+const _ = require('lodash');
+
+// Specify all the files and directories a template can have
+const allowChildren = '*';
+const allowedTemplateTree = {
+  // Root template files
+  'template.json': true,
+  'README.md': true,
+  '.gitignore': true,
+  // Template contents
+  template: {
+    'README.md': true,
+    '.env.example': true,
+    api: allowChildren,
+    components: allowChildren,
+    config: {
+      functions: allowChildren,
+    },
+    data: allowChildren,
+    plugins: allowChildren,
+    public: allowChildren,
+    scripts: allowChildren,
+  },
+};
+
+// Traverse template tree to make sure each file and folder is allowed
+async function checkTemplateStructure(templatePath) {
+  // Recursively check if each item in the template is allowed
+  const checkPathContents = (pathToCheck, parents) => {
+    const contents = fse.readdirSync(pathToCheck);
+    contents.forEach(item => {
+      const nextParents = [...parents, item];
+      const matchingTreeValue = _.get(allowedTemplateTree, nextParents);
+
+      // Treat files and directories separately
+      const itemPath = path.resolve(pathToCheck, item);
+      const isDirectory = fse.statSync(itemPath).isDirectory();
+
+      if (matchingTreeValue === undefined) {
+        // Unknown paths are forbidden
+        throw Error(`Illegal template structure, unknown path ${nextParents.join('/')}`);
+      }
+
+      if (matchingTreeValue === true) {
+        if (!isDirectory) {
+          // All good, the file is allowed
+          return;
+        }
+        throw Error(
+          `Illegal template structure, expected a file and got a directory at ${nextParents.join(
+            '/'
+          )}`
+        );
+      }
+
+      if (isDirectory) {
+        if (matchingTreeValue === allowChildren) {
+          // All children are allowed
+          return;
+        }
+        // Check if the contents of the directory are allowed
+        checkPathContents(itemPath, nextParents);
+      } else {
+        throw Error(`Illegal template structure, unknow file ${nextParents.join('/')}`);
+      }
+    });
+  };
+
+  checkPathContents(templatePath, []);
+}
+
+function getRepoInfo(githubURL) {
+  // Make sure it's a github url
```

Comment:
You might want to consider using https://github.com/npm/hosted-git-info to be safe in parsing the url